### PR TITLE
Linux amd64 abi fixes

### PR DESCRIPTION
**Pull request by arnetheduck**
This builds on #5656 to fix abi on linux amd64. Instead of going the way of #5018 which touched all arches, this branch isolates the changes in specific linux_amd64 blocks - this is the similar to how for example the musl c library approaches cross-platform.

A natural extension here would be to automate the process, but since that's far off, this branch brute-forces it ;) Let's see if the tests are happy.

### CODE REVIEW

File: tools/detect/detect.nim [hunk 1]
```diff
@@ -223,19 +223,6 @@ if header("<fcntl.h>"):
   v("POSIX_FADV_DONTNEED")
   v("POSIX_FADV_NOREUSE")
 
-if header("<fenv.h>"):
```
Comment (Araq) [inline tools/detect/detect.nim]: Did you use the ``detect`` tool or why did you update it?

File: lib/system/threads.nim [hunk 2]
```diff
@@ -177,18 +177,28 @@ else:
     else:
       type Time = int
 
+  when defined(linux) and defined(amd64):
+    type
+      SysThread* {.importc: "pthread_t", header: "<sys/types.h>" .} = distinct culong
+      Pthread_attr {.importc: "pthread_attr_t",
+                       header: "<sys/types.h>", final, pure.} = object
+        abi: array[56, uint8]
```
Comment (Araq) [inline lib/system/threads.nim]: I'm sure Pthread_attr needs alignment beyond uint8, so you need to use ``array[7, float64]`` or something.

File: lib/system/syslocks.nim [hunk 3]
```diff
@@ -102,12 +102,21 @@ else:
     SysLock {.importc: "pthread_mutex_t", pure, final,
                header: """#include <sys/types.h>
                           #include <pthread.h>""".} = object
+      when defined(linux) and defined(amd64):
+        abi: array[40, uint8]
```
Comment (Araq) [inline lib/system/syslocks.nim]: alignment?

File: lib/system/syslocks.nim [hunk 4]
```diff
@@ -102,12 +102,21 @@ else:
...
+
     SysLockAttr {.importc: "pthread_mutexattr_t", pure, final
                header: """#include <sys/types.h>
                           #include <pthread.h>""".} = object
+      when defined(linux) and defined(amd64):
+        abi: array[4, uint8]  # actually a cint
```
Comment (Araq) [inline lib/system/syslocks.nim]: alignment?

File: lib/system/syslocks.nim [hunk 5]
```diff
@@ -102,12 +102,21 @@ else:
...
+
     SysCond {.importc: "pthread_cond_t", pure, final,
                header: """#include <sys/types.h>
                           #include <pthread.h>""".} = object
+      when defined(linux) and defined(amd64):
+        abi: array[48, uint8]
```
Comment (Araq) [inline lib/system/syslocks.nim]: alignment?

File: lib/system/ansi_c.nim [hunk 6]
```diff
@@ -26,8 +26,14 @@ proc c_memset(p: pointer, value: cint, size: csize): pointer {.
 proc c_strcmp(a, b: cstring): cint {.
   importc: "strcmp", header: "<string.h>", noSideEffect.}
 
-type
-  C_JmpBuf {.importc: "jmp_buf", header: "<setjmp.h>".} = object
+when defined(linux) and defined(amd64):
+  type
+    C_JmpBuf {.importc: "jmp_buf", header: "<setjmp.h>", pure, final, bycopy.} = object
+        abi: array[200, uint8]
```
Comment (Araq) [inline lib/system/ansi_c.nim]: alignment?

File: lib/pure/times.nim [hunk 7]
```diff
@@ -1116,7 +1127,7 @@ when not defined(JS):
         gmtoff {.importc: "tm_gmtoff".}: clong
   else:
     type
-      StructTM {.importc: "struct tm", final.} = object
+      StructTM {.importc: "struct tm", pure, final.} = object
```
Comment (Araq) [inline lib/pure/times.nim]: the ``pure`` here is pointless.

File: lib/pure/times.nim [hunk 8]
```diff
@@ -1103,7 +1114,7 @@ when not defined(JS):
   when defined(freebsd) or defined(netbsd) or defined(openbsd) or
       defined(macosx):
     type
-      StructTM {.importc: "struct tm", final.} = object
+      StructTM {.importc: "struct tm", pure, final.} = object
```
Comment (Araq) [inline lib/pure/times.nim]: the ``pure`` here is pointless

**Comment by Araq** [comment]
Apart from my other remarks, the duplication of ``posix.nim`` is not acceptable. Please come up with something like 

```nim

template posixconst(name, val: untyped, typ = cint, underscore = false) =
  when defined(linux) and defined(amd64):
    const name = typ(val)
  else:
    var name {.importc: (when underscore: "_" else: " ") & astToStr(name).}: typ
```

**Comment by arnetheduck** [comment]
normally, I'm not a fan of duplication, but I'd make an exception in this case:
* the values and functions in it will _never_ change or be removed once they are correct (stuff might get added), which makes maintenance much less of a problem, and therefore lowers the cost of duplication. in this branch, I fixed the types but not the function declarations - they're wrong too, but that's less bad
* the suggested template doesn't scale for multiple arches, specially when not all arches have the same symbols available
* once more platforms have been fixed to be correct, it's possible to start merging back the files using a hierarchical approach - ie linux x86 & x86_64 share a lot of definitions so it's possible to maintain a fairly simple hierarchy and get rid of some duplication, but I'm not sure it's worth the effort
* the files become extremely easy to read & process, and given the large number of symbols, it might be easier on the compiler, instead of having to instantiate a nasty template for each one
* being easy to read, and not hidden behind a jungle of whens and templates is an advantage when trying to debug anything abi related
* splitting makes it easier to autogenerate (parts) in the future because typically, the tool can only run on one platform at a time

File: lib/system/threads.nim (hunk 2 above)
Comment (arnetheduck) [inline lib/system/threads.nim]: ah yes, that's a clever way to do it, was gonna look into it next

File: lib/system/syslocks.nim (hunk 3 above)
Comment (arnetheduck) [inline lib/system/syslocks.nim]: fixed

File: lib/system/syslocks.nim (hunk 4 above)
Comment (arnetheduck) [inline lib/system/syslocks.nim]: fixed

File: lib/system/syslocks.nim (hunk 5 above)
Comment (arnetheduck) [inline lib/system/syslocks.nim]: fixed

**Comment by Araq** [comment]
> splitting makes it easier to autogenerate (parts) in the future because typically, the tool can only run on one platform at a time

That's what the detect tool tries to do. So either generate it (this means getting the generator into a usable state) or use code generation via templates/macros. This manual duplication of code **will not** be accepted. Sorry.

**Comment by Araq** [comment]
Also I think you should split this PR into 2: One for the minor things that will be accepted as they are and the other for the posix.nim refactoring.

**Comment by arnetheduck** [comment]
>      That's what the detect tool tries to do. So either generate it (this means getting the generator into a usable state) or use code generation via templates/macros as I suggested. This manual duplication of code will not be accepted. Sorry.

The constants were in fact generated by the detect tool and simply pasted into the file, so is that sufficiently automatic? I could make detect write the `var` part of the generic posix.nim too and someone with access can easily run it for other platforms. That works as long as one accepts that posix.nim, the generic one, will be the union of all constants across all platforms, even ones that might not exist for a particular platform (ie the nim part will compile, but C compiler / linker will complain).

Another thing with the current constants is that they're (almost) all declared as cint - this is wrong for a lot of them, but it's.. tricky.. to do it right and readable. `detect.nim` can do values, but not type, and if type is different across platforms, it's gonna be tricky without some duplication.

Specifically, I don't see how I can make a template or macro that will be readable enough that it can practically be used - the suggested one kind of works for 1 platform, but when more are added, there seems to be few reasonable ways of making them viewable and editable - just adding a param per platform doesn't scale and doesn't address partial support.

What about type declarations? When jungle, or is split-files with some repetition ok here? Eventually, I'd like to autogen these header by header using c2nim, but unfortunately c2nim can't parse glibc headers. I have a plan, but it might take a while.

A horrible example of when jungle is `system.nim` where it's near impossible to tell which code actually gets run (I've even hacked the compiler to print it out, much like running the preprocessor in C, which feels like a step backwards and in general, lack of design or simply something that needs a refactoring round to clean up the cruft).

Here's what it would look like with when's - note that there's not really less duplication here than there would be when splitting files - you've just managed to duplicate the when's instead of duplicating the type names (and yes, it's really that bad - for system types, order, size and presence of fields changes quite a bit, even between say.. linux 32 and 64 bits, or linux for arm and linux for intel):

    type A
      when defined(foo) and defined(bar):
         b: int
         c: int
      elif defined(xxx):
         c: int
         b: int
      else
        ...

And lastly, the proc declarations, ok to have multiple files per platform for these? They sit somewhere between consts and types, in that variation is less than types but  availability and param/ret types are just as varied (and no templace can save us here). They are however more amenable to c2nim-like solutions.

**Comment by arnetheduck** [comment]
Which parts are ok for inclusion now? the types in system and friends (that ironically duplicate posix! I think there's like 5 different definitions of `Timeval` around, with slightly different field types)? 

The compiler changes needed are already separately available as  #5656 - it's one line.

File: lib/pure/times.nim (hunk 7 above)
Comment (arnetheduck) [inline lib/pure/times.nim]: added it mainly for consistency. there's lots of `pure` everywhere, in fact most `importc` have them.. should they all be removed instead? does `importc` imply `pure`? the manual is somewhat unclear here - "they're no longer needed for abi compatibility" - but if I create a random Nim type without `importc` and without `pure`, it will not be abi compatible with c - it will have a type header.. for example, if I'm developing a shared memory transport, I cannot simply memcpy c types, and do the same in nim without pure, and expect it to work.

Comment (Araq) [inline lib/pure/times.nim]: >  but if I create a random Nim type without importc and without pure, it will not be abi compatible with c - it will have a type header..

It really shouldn't have a type header. Are you sure?

**Comment by Araq** [comment]
> The constants were in fact generated by the detect tool and simply pasted into the file, so is that sufficiently automatic?

No. It should generate a whole include file ``linux_amd64_consts.nim``.

**Comment by Araq** [comment]
> Which parts are ok for inclusion now? the types in system and friends ...?

Exactly.

File: lib/pure/times.nim (hunk 7 above)
Comment (arnetheduck) [inline lib/pure/times.nim]: oops, looks like I was wrong. was looking at an inheritable object which obviously cannot be used across a c abi. by the same vein, `final` looks redundant - it's set implicitly - the manuals really hard to follow on these subtle points.

**Comment by arnetheduck** [comment]
Minor refactoring in #5692, rest will come later, separately
